### PR TITLE
Fix #323 - Catch lexical_cast errors when reading XML

### LINKED ISSUE #323: Bad error message when attribute can't convert to double (Bugzilla #1230)
On 2013-04-19 16:38:13, @macumber wrote:

> Improve that error message so it tells us the name of the attribute and prints its value as a string.

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes #323 
 - Catches it in BCLComponent.
 - The Attribute class itself already handles it: https://github.com/NREL/OpenStudio/blob/develop/src/utilities/data/Attribute.cpp#L228-L234

### Pull Request Author

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Checked behavior in OpenStudioApplication, adjusted policies as needed (`src/openstudio_lib/library/OpenStudioPolicy.xml`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: resources/utilities/BCL/Components/323_wrongXMLType_float/component.xml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  <attribute>
+   <name>Standard</name>
+   <value>ASHRAE Std189</value>
+   <datatype>string</datatype>
+  </attribute>
+  <attribute>
+   <name>Effective R-value</name>
+   <value>THIS IS A MISTAKE</value>
+   <datatype>float</datatype>
+  </attribute>
```
Comment (jmarrec) [inline resources/utilities/BCL/Components/323_wrongXMLType_float/component.xml:23]: Test one: for float

File: resources/utilities/BCL/Components/323_wrongXMLType_int/component.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  <attribute>
+   <name>Standard</name>
+   <value>ASHRAE Std189</value>
+   <datatype>string</datatype>
+  </attribute>
+  <attribute>
+   <name>Effective R-value</name>
+   <value>THIS IS A MISTAKE</value>
+   <datatype>int</datatype>
+  </attribute>
```
Comment (jmarrec) [inline resources/utilities/BCL/Components/323_wrongXMLType_int/component.xml:23]: Test one: for int

File: src/utilities/bcl/test/BCLComponent_GTest.cpp [hunk 3]
```diff
@@ -41,3 +43,31 @@ TEST_F(BCLFixture, BCLComponent) {
...
+
+  StringStreamLogSink sink;
+  sink.setLogLevel(Error);
+
+  EXPECT_THROW(BCLComponent{compDir}, openstudio::Exception);
+  ASSERT_EQ(1, sink.logMessages().size());
+  EXPECT_EQ("Please double check your XML, you have an attribute named 'Effective R-value' with a datatype of 'int' but the value isn't an integer: "
+            "'THIS IS A MISTAKE'.",
+            sink.logMessages()[0].logMessage());
+}
```
Comment (jmarrec) [inline src/utilities/bcl/test/BCLComponent_GTest.cpp:73]: Do the actual Gtest

File: src/utilities/bcl/BCLComponent.cpp [hunk 4]
```diff
@@ -76,21 +76,32 @@ BCLComponent::BCLComponent(const openstudio::path& dir) : m_directory(dir) {
         std::string units = componentElement.child("units").text().as_string();
 
         if (datatype == "float") {
+          double castVal;
+          try {
+            castVal = boost::lexical_cast<double>(value);
+          } catch (const boost::bad_lexical_cast&) {
+            LOG_AND_THROW("Please double check your XML, you have an attribute named '" << name << "' with a datatype of '" << datatype
+                                                                                        << "' but the value isn't a double: '" << value << "'.");
+          }
```
Comment (jmarrec) [inline src/utilities/bcl/BCLComponent.cpp:85]: Catch a failure in lexical_cast, and issue an informative message

File: src/utilities/bcl/BCLComponent.cpp [hunk 5]
```diff
@@ -76,21 +76,32 @@ BCLComponent::BCLComponent(const openstudio::path& dir) : m_directory(dir) {
...
           if (units.empty()) {
-            Attribute attr(name, boost::lexical_cast<double>(value));
-            m_attributes.push_back(attr);
+            m_attributes.emplace_back(name, castVal);
```
Comment (jmarrec) [inline src/utilities/bcl/BCLComponent.cpp:87]: While I'm at it, avoid unecessary copy/move and create object directly in vector via emplace_back

File: src/utilities/bcl/BCLComponent.cpp [hunk 6]
```diff
@@ -76,21 +76,32 @@ BCLComponent::BCLComponent(const openstudio::path& dir) : m_directory(dir) {
...
+            m_attributes.emplace_back(name, castVal, units);
           }
         } else if (datatype == "int") {
+          int castVal;
+          try {
+            castVal = boost::lexical_cast<int>(value);
+          } catch (const boost::bad_lexical_cast&) {
+            LOG_AND_THROW("Please double check your XML, you have an attribute named '" << name << "' with a datatype of '" << datatype
+                                                                                        << "' but the value isn't an integer: '" << value << "'.");
+          }
```
Comment (jmarrec) [inline src/utilities/bcl/BCLComponent.cpp:98]: Same for int

**Review by tijcolem — APPROVED** [review]

**Pull request merged**